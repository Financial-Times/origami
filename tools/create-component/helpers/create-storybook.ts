import * as jetPack from 'fs-jetpack';
export async function createStoryBookBoilerPlate(toolbox) {
	const {prompt, print} = toolbox;
	const props = await prompt.ask(componentName);
	console.log(props.name);
	print.warning(`ok! generating "${props.name}" in components folder!`);
	await copyTemplates(toolbox, props);
	print.success(`yay! "${props.name}" is ready!`);
}

const componentName = {
	type: 'autocomplete',
	name: 'name',
	message: 'Choose component:',
	limit: 10,
	initial: 2,
	choices: findComponentsWithoutStoryBook(),
};

function findComponentsWithoutStoryBook() {
	const components = jetPack
		.cwd('../../components')
		.find({
			matching: '*',
			directories: true,
			recursive: false,
			files: false,
		})
		.filter(dir => !jetPack.exists(`${dir}/stories`))
		.map(dir => dir.replace('../../components/', ''));
	return components;
}

async function copyTemplates(toolbox, props) {
	const storyBookFiles = jetPack
		.find('src/templates/stories')
		.map(file => file.split('templates/')[1]);
	const tsx = jetPack
		.find('src/templates/src/tsx')
		.map(file => file.split('templates/')[1]);
	const files = [...storyBookFiles, ...tsx];
	await Promise.all(
		files.map(async file => {
			const template = `/${file}`;
			// Where to copy this file to.
			const target = `../../components/${props.name}/${file
				.replace('<name>', trimName(props.name))
				.replace('.ejs', '')}`;
			return await toolbox.template.generate({
				template,
				target,
				props: {...props},
			});
		})
	);
}

function trimName(name) {
	return name.replace('o-', '');
}
