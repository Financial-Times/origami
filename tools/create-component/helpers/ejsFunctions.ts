import * as jetPack from 'fs-jetpack';
import * as ejs from 'ejs';
import {PropsType} from '../src/types';
import {
	oComponentsPath,
	templateFiles,
	jsFilesNotToInclude,
	sassFilesNotToInclude,
} from './utils';

export async function copyTemplates(
	props: PropsType,
	targetPath: string = oComponentsPath
): Promise<void> {
	let files = templateFiles;
	if (!props.javascript) {
		files = files.filter(file => !jsFilesNotToInclude.includes(file));
	}

	if (!props.scss) {
		files = files.filter(file => !sassFilesNotToInclude.includes(file));
	}
	await Promise.all(
		files.map(async file => {
			const templateStr = await ejsToHTML(file, {
				props: {...props},
			});
			const target = `${targetPath}/${props.name}/${file
				.replace('<name>', props.name)
				.replace('.ejs', '')}`;
			HtmlStringToFile(target, templateStr);
		})
	);
}

export const ejsToHTML = async (
	file: string,
	data?: {props: PropsType}
): Promise<string> => {
	const filePath = jetPack.path('src', 'templates', file);
	const template = jetPack.read(filePath);
	data ? (data.props.lowercase = lowercase) : null;
	data ? (data.props.titleCase = titleCase) : null;
	data ? (data.props.camelCase = camelCase) : null;
	data ? (data.props.withoutPrefix = withoutPrefix) : null;
	const str = await ejs.render(template, data, {
		async: true,
	});
	return str;
};

export const HtmlStringToFile = (targetLocation: string, str: string): void => {
	jetPack.write(targetLocation, str);
};

const lowercase = string => string.toLowerCase();
const titleCase = string =>
	string
		.split('-')
		.map(word => word.replace(word[0], word[0].toUpperCase()))
		.join('');
const camelCase = string =>
	string.replace(/\-+(.)/g, (match, chr) => chr.toUpperCase());
const withoutPrefix = string => string.replace(/^[a-z]-/g, '');
