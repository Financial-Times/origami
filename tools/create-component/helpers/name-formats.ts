export const lowercase = (string) => string.toLowerCase()
export const titleCase = (string) =>
	string
		.split('-')
		.map((word) => word.replace(word[0], word[0].toUpperCase()))
		.join('')
export const camelCase = (string) =>
	string.replace(/\-+(.)/g, (match, chr) => chr.toUpperCase())
export const withoutPrefix = (string) => string.replace(/^[a-z]-/g, '')
