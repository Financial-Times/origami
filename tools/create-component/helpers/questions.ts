export const componentName = {
	type: 'input',
	name: 'name',
	message: 'component name (required)',
	validate: (value) => {
		if (!value) {
			return 'please enter a name for your component\nhttps://origami.ft.com/docs/components/code/#naming-conventions'
		}
		return true
	},
}

const description = {
	name: 'description',
	type: 'input',
	message: 'description (required)',
	validate: (value) => {
		if (!value) {
			return 'please enter a description for your component\nhttps://origami.ft.com/docs/manifests/origami-json/#description'
		}
		return true
	},
}
const keywords = {
	name: 'keywords',
	type: 'input',
	message: 'keywords (to help discoverability in the registry)',
}

const form = {
	name: 'team',
	type: 'form',
	message: 'Please provide the following information:',
	choices: [
		{
			name: 'email',
			message: 'Support team email (required)',
			initial: 'origami.support@ft.com',
		},
		{
			name: 'slack',
			message: 'Support team slack channel (required)',
			initial: '#origami-support',
		},
		{
			name: 'githubTeam',
			message: 'Support Github team (required)',
			initial: 'origami-core',
		},
	],
}

const status = {
	type: 'select',
	name: 'status',
	message:
		'support status (https://origami.ft.com/docs/manifests/origami-json/#supportstatus)',
	initial: 'experimental',
	choices: ['active', 'experimental', 'maintained'],
}
const brands = {
	type: 'multiselect',
	name: 'brands',
	message: 'choose templates for files:',
	initial: ['core'],
	choices: ['core', 'internal', 'whitelabel'],
}
const javascript = {
	name: 'javascript',
	type: 'confirm',
	message:
		'will your component use JavaScript? (this will add some helpful Origami JavaScript boilerplate)',
	initial: true,
}
const scss = {
	name: 'scss',
	type: 'confirm',
	message:
		'will your component use Sass? (this will add some helpful Origami Sass boilerplate)',
	initial: true,
}
export const questions = [
	description,
	keywords,
	form,
	status,
	brands,
	javascript,
	scss,
]
