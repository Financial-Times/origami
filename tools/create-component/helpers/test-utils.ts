export const keys = {
	up: '\x1B\x5B\x41',
	down: '\x1B\x5B\x42',
	enter: '\x0D',
	space: '\x20',
};
// helper function for timing
export const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
