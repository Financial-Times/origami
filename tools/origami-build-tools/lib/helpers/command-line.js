'use strict';

const execa = require('execa');
const process = require('process');

function run(command, args, options) {
	options = options || {};
	const opts = {
		cwd: options.cwd || process.cwd(),
		preferLocal: true
	};
	if (options.localDir) {
		opts.localDir = options.localDir;
	}
	if (options && options.stdout === undefined && options.stderr === undefined) {
		opts.stdio = 'inherit';
	}
	const result = execa(command, args, opts);
	if (options && options.stdout === undefined) {
		options.stdout = process.stdout;
	}
	if (options && options.stderr === undefined) {
		options.stderr = process.stderr;
	}
	if (options.stdout && !opts.stdio) {
		result.stdout.pipe(options.stdout);
	}
	if (options.stderr && !opts.stdio) {
		result.stderr.pipe(options.stderr);
	}
	return result;
}

module.exports.run = run;
