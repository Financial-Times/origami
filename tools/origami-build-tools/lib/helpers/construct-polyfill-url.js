'use strict';

const path = require('path');
const globby = require('globby');
const { readFile } = require('fs/promises');

async function constructBrowserDeps() {
	const dependencies = await globby(['node_modules/*/*/origami.json', 'node_modules/*/origami.json', 'origami.json']);
	const requiredFeatures = [];
	await Promise.all(dependencies.map(async dependency => {
		if (dependency.startsWith('/')) {
			dependency = dependency.substr(1);
		}
		const origami = JSON.parse(await readFile(path.resolve(dependency), 'utf-8'));
		if (origami.browserFeatures && origami.browserFeatures.required) {
			requiredFeatures.push(...origami.browserFeatures.required);
		}
	}));

	const features = Array.from(new Set(requiredFeatures));
	// Include the `CustomEvent` polyfill so that demos which include dev
	// dependencies work in older browsers. This is because the Origami Build
	// Service includes `o-autoinit` which requires the `CustomElement` polyfill.
	// https://github.com/Financial-Times/origami-build-tools/issues/906
	features.push('CustomEvent');
	if (features.length > 0) {
		return `https://polyfill.io/v3/polyfill.min.js?features=${features.join(',')}&flags=gated&unknown=polyfill`;
	} else {
		return 'https://polyfill.io/v3/polyfill.min.js?flags=gated&unknown=polyfill';
	}
}

module.exports = constructBrowserDeps;
