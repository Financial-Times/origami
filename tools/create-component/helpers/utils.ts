import { filesystem } from 'gluegun'
import { componentName, questions } from './questions'

export const oComponentsPath = filesystem.path(__dirname, '../../../components')
export const templateFiles = filesystem
	.find('src/templates')
	.map((file) => file.split('templates/')[1])

export async function getComponentName(toolbox) {
	const { prompt, print } = toolbox
	const nameAnswer = await prompt.ask(componentName)
	const sanitized = sanitizeName(nameAnswer.name)
	const sanityCheck = await prompt.confirm(`Is "${sanitized}" okay?`, true)

	if (!sanityCheck) {
		return getComponentName(toolbox)
	}
	if (findExistingComponents().includes(sanitized)) {
		print.error(`Component "${sanitized}" already exists.`)
		return getComponentName(toolbox)
	}

	return sanitized
}
export function sanitizeName(name) {
	return name
		.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`) // lowers capitalized letters, adds '-' before each (newComponent => new-component)
		.toLowerCase()
		.replace(/([^a-z-])/g, '-') // replaces any characters that are not a-z or a '-' with a hyphen
		.replace(/(-{2,})/g, '-') // replaces double hyphen with single hyphen
		.replace(/(^-|-$)/g, '') // replaces hyphen if at the beginning or end of string
}

export async function getFinalConfirmation(toolbox, props) {
	const { prompt, print } = toolbox
	print.debug(props, 'Component summary')

	const finalConfirm = await prompt.confirm(
		'Do you want to create this component?',
		true
	)
	if (finalConfirm) {
		return props
	}
	const changeConfirm = await prompt.confirm(
		'Do you want to change the component details?',
		false
	)
	if (!finalConfirm && changeConfirm) {
		const responses = await prompt.ask({
			type: 'multiselect',
			name: 'change',
			message: 'What would you like to change?',
			choices: Object.keys(props),
		})
		let name = props.name
		let keywords = props.keywords

		if (responses.change.includes('name')) {
			name = await getComponentName(toolbox)
		}

		const newAnswers = await prompt.ask(
			questions.filter((q) => responses.change.includes(q.name))
		)

		if (responses.change.includes('keywords')) {
			keywords = generateKeywords(newAnswers.keywords)
		}

		const changedAnswers = {
			...props,
			name,
			...newAnswers,
			keywords,
		}

		return await getFinalConfirmation(toolbox, changedAnswers)
	}
}

function findExistingComponents() {
	return filesystem.list(oComponentsPath)
}

export function generateKeywords(input) {
	return [...new Set(input.split(/\s*[\s,]\s*/).filter(Boolean))]
}
