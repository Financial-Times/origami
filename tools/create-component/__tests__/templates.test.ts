import {ejsToHTML} from '../helpers/ejsFunctions';

describe('Component template files', () => {
	const data = {
		props: {
			name: 'o-test-name',
			description: 'bla blas ',
			keywords: ['git', 'o', 'init'],
			team: {
				email: 'origami.support@ft.com',
				slack: '#origami-support',
				githubTeam: 'origami-core',
			},
			status: 'experimental',
			brands: ['core', 'internal'],
			javascript: true,
			scss: true,
		},
	};
	describe('Root non ejs files', () => {
		it('expect .eslintignore ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.eslintignore');
			expect(str).toMatchSnapshot();
		});
		it('expect .eslintrc.cjs ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.eslintrc.cjs');
			expect(str).toMatchSnapshot();
		});
		it('expect .gitignore ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.gitignore');
			expect(str).toMatchSnapshot();
		});
		it('expect .npmignore ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.npmignore');
			expect(str).toMatchSnapshot();
		});
		it('expect .remarkrc.cjs ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.remarkrc.cjs');
			expect(str).toMatchSnapshot();
		});
		it('expect .stylelintignore ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.stylelintignore');
			expect(str).toMatchSnapshot();
		});
		it('expect .stylelintrc.cjs ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('.stylelintrc.cjs');
			expect(str).toMatchSnapshot();
		});
	});

	describe('Root ejs files', () => {
		it('expect main.js ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('main.js.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect main.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('main.scss.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect origami.json ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('origami.json.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect package.json ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('package.json.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect README.md ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('README.md.ejs', data);
			expect(str).toMatchSnapshot();
		});
	});

	describe('demos ejs files', () => {
		it('expect demo.js ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('demos/src/demo.js');
			expect(str).toMatchSnapshot();
		});
		it('expect demo.mustache ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('demos/src/demo.mustache.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect demo.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('demos/src/demo.scss.ejs', data);
			expect(str).toMatchSnapshot();
		});
	});
	describe('src and stories ejs files', () => {
		it(`expect <name>.js ejsRenderer fn output to match snapshot`, async () => {
			const str = await ejsToHTML(`src/js/<name>.js.ejs`, data);
			expect(str).toMatchSnapshot();
		});
		it('expect _brand.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('src/scss/_brand.scss.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect _variables.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('src/scss/_variables.scss.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it(`expect <name>.tsx ejsRenderer fn output to match snapshot`, async () => {
			const str = await ejsToHTML(`src/tsx/<name>.tsx.ejs`, data);
			expect(str).toMatchSnapshot();
		});
		it(`expect <name>.tsx ejsRenderer fn output to match snapshot`, async () => {
			const str = await ejsToHTML(`stories/<name>.scss.ejs`, data);
			expect(str).toMatchSnapshot();
		});
		it(`expect <name>.stories.tsx ejsRenderer fn output to match snapshot`, async () => {
			const str = await ejsToHTML(`stories/<name>.stories.tsx.ejs`, data);
			expect(str).toMatchSnapshot();
		});
	});

	describe('test ejs files', () => {
		it(`expect <name>.test.js ejsRenderer fn output to match snapshot`, async () => {
			const str = await ejsToHTML(`test/<name>.test.js.ejs`, data);
			expect(str).toMatchSnapshot();
		});
		it('expect fixtures.js ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('test/helpers/fixtures.js.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect _main.test.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('test/scss/_main.test.scss.ejs', data);
			expect(str).toMatchSnapshot();
		});
		it('expect index.test.scss ejsRenderer fn output to match snapshot', async () => {
			const str = await ejsToHTML('test/scss/index.test.scss');
			expect(str).toMatchSnapshot();
		});
	});
});
