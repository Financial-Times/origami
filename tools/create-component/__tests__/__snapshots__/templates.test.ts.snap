// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component template files Root ejs files expect README.md ejsRenderer fn output to match snapshot 1`] = `
"# o-test-name

_A short description of what the component does._

- [Usage](#usage)
- [Behaviour](#behaviour)
- [Markup](#markup)
- [Sass](#sass)
- [JavaScript](#javascript)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)
- [Migration](#migration)
- [Contact](#contact)
- [Licence](#licence)

_Instructions for the component. We've broken this down by Markup, JavaScript, Sass, etc. You may add more sections as needed._

## Usage

Check out [how to include Origami components in your project](https://origami.ft.com/docs/components/#including-components-in-your-project) to get started with \`o-test-name\`.

## Behaviour

\`o-test-name\` Behaviour description.

## Markup

_Markup documentation for the component. Including css classes, data attributes, and tags for variations of the component._
_Demo markup is also shown in the registry. For complex markup it can be useful to describe the markup and link to the registry demos for a full example._
_Remember to start your codeblocks with three backticks and \\"html\\" so your markup is syntax highlighted correctly._
\`\`\`html
<div data-o-component=\\"o-test-name\\" class='o-test-name'>
</div>
\`\`\`

## Sass

_Sass documentation. With instructions to include styles with a primary mixin, what options are available, brand support where applicable._
_For complex components it may be helpful to document apis with SassDoc and link to the components SassDoc in the Origami Registry._
_Remember to start your codeblocks with three backticks and \\"scss\\" so your markup is syntax highlighted correctly._

Use \`@include oTestName()\` to include styles for all \`o-test-name\` features.

\`\`\`scss
@import \\"@financial-times/o-test-name\\";

@include oTestName();
\`\`\`


## JavaScript

_JavaScript documentation. Including how to initialise the component, available options, and common usecases._
_For complex components it may be helpful to document apis with JSDoc and link to the components JSDocs in the Origami Registry._
_Remember to start your codeblocks with three backticks and \\"js\\" so your js is syntax highlighted correctly._

JavaScript is initialised automatically for [Origami Build Service](https://www.ft.com/__origami/service/build/v2/) users. If your project is using a manual build process, [initialise  \`o-test-name\` manually](https://origami.ft.com/docs/tutorials/manual-build/).

For example call the \`init\` method to initialise all \`o-test-name\` instances in the document:

\`\`\`js
import oTestName from 'o-test-name';
oTestName.init();
\`\`\`

Or pass an element to initialise a specific \`o-test-name\` instance:

\`\`\`js
import oTestName from 'o-test-name';
const oTestNameElement = document.getElementById('#my-o-test-name-element');
oTestName.init(oTestNameElement);
\`\`\`



## Troubleshooting

_Answers to questions or problems that come up repeatedly._
_It's likely you can delete this section for now. Remember to remove it from the table of contents._

## Contributing

_Contributing documentation for non typical components._
_It's likely you can delete this section if your component conforms with other components._
_Remember to update the table of contents is you remove this section._

## Migration

_Guides to upgrade from one major version of a component to another._
_This section includes a table of major versions with links to \`MIGRATION.md\`._
_\`MIGRATION.md\` would provide upgrade instructions, but since this is a new component there is no \`MIGRATION.md\`._
_Delete this section, including from the table of contents, until there is a second major version of this component._

State | Major Version | Last Minor Release | Migration guide |
:---: | :---: | :---: | :---:
✨ active | 3 | N/A | [migrate to v3](MIGRATION.md#migrating-from-v2-to-v3) |
⚠ maintained | 2 | 2.0 | [migrate to v2](MIGRATION.md#migrating-from-v1-to-v2) |
╳ deprecated | 1 | 1.0 | N/A |

## Contact
If you have any questions or comments about this component, or need help using it, please either [raise an issue](https://github.com/Financial-Times/origami/issues/new?labels=o-test-name,components), visit [#origami-support](https://financialtimes.slack.com/messages/#origami-support/) or email [origami.support@ft.com](mailto:origami.support@ft.com).

## Licence
This software is published by the Financial Times under the [MIT licence](http://opensource.org/licenses/MIT).
\`;
};
"
`;

exports[`Component template files Root ejs files expect main.js ejsRenderer fn output to match snapshot 1`] = `
"import TestName from './src/js/test-name.js';

const constructAll = () => {
	TestName.init();
	document.removeEventListener('o.DOMContentLoaded', constructAll);
};

if (typeof document !== 'undefined') {
	document.addEventListener('o.DOMContentLoaded', constructAll);
}

export default TestName;
"
`;

exports[`Component template files Root ejs files expect main.scss ejsRenderer fn output to match snapshot 1`] = `
"@import '@financial-times/o-brand/main';

@import './src/scss/variables';
@import 'src/scss/brand';
/// Output all oTestName features
/// @param {Map} $opts [()] - A map of options to configure the output
/// @access public
/// @example scss
///		@include oTestName($opts: (
///			// your opts here
///		))
@mixin oTestName ($opts: ()) {
	// content of primary mixin
	.o-test-name {
		display: block;
	}
}

"
`;

exports[`Component template files Root ejs files expect origami.json ejsRenderer fn output to match snapshot 1`] = `
"{
	\\"origamiType\\": \\"component\\",
	\\"origamiVersion\\": \\"2.0\\",
	\\"brands\\": [\\"core\\",\\"internal\\"],
	\\"support\\": \\"https://github.com/Financial-Times/origami/issues/new?labels=o-test-name,components\\",
	\\"supportContact\\": {
		\\"email\\":  \\"origami.support@ft.com\\",
		\\"slack\\": \\"#origami-support\\"
	},
	\\"supportStatus\\": \\"experimental\\",
	\\"browserFeatures\\": {},
	\\"demosDefaults\\": {
		\\"sass\\": \\"demos/src/demo.scss\\", 
		\\"js\\": \\"demos/src/demo.js\\", 
		\\"template\\": \\"demos/src/demo.mustache\\",
		\\"documentClasses\\": \\"\\",
		\\"dependencies\\": []
	},
	\\"demos\\": [
		{
			\\"title\\": \\"A Useful Demo\\",
			\\"name\\": \\"demo\\",
			\\"template\\": \\"demos/src/demo.mustache\\",
			\\"description\\": \\"Description of the demo\\"
		}
	]
}
"
`;

exports[`Component template files Root ejs files expect package.json ejsRenderer fn output to match snapshot 1`] = `
"{
  \\"name\\": \\"@financial-times/o-test-name\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"bla blas \\",
  \\"keywords\\": [\\"git\\",\\"o\\",\\"init\\"],
  \\"homepage\\": \\"https://registry.origami.ft.com/components/o-test-name\\",
  \\"bugs\\": {
    \\"url\\": \\"https://github.com/Financial-Times/origami/issues/new?labels=o-test-name,components\\",
    \\"email\\": \\"origami.support@ft.com\\"
  },
  \\"license\\": \\"MIT\\",
  \\"type\\": \\"module\\",
  \\"browser\\": \\"main.js\\",
  \\"scripts\\": {
    \\"start\\": \\"npx serve ./demos/local\\",
    \\"build\\": \\"bash ../../scripts/component/build.bash\\",
    \\"test\\": \\"bash ../../scripts/component/test.bash\\",
    \\"debug:js\\": \\"bash ../../scripts/component/debug-js.bash\\",
    \\"lint\\": \\"bash ../../scripts/component/lint.bash\\",
    \\"watch\\": \\"bash ../../scripts/component/watch.bash\\"
  },
	\\"peerDependencies\\": {
			\\"@financial-times/o-brand\\": \\"^4.1.0\\"
	},
	  \\"devDependencies\\": {},
  \\"engines\\": {
    \\"npm\\": \\"^7 || ^8\\"
  },
  \\"volta\\": {
    \\"node\\": \\"14.16.1\\",
    \\"npm\\": \\"7.11.1\\"
  },
  \\"percy\\": true,
  \\"private\\": false
}
"
`;

exports[`Component template files Root non ejs files expect .eslintignore ejsRenderer fn output to match snapshot 1`] = `
"demos/local/**
"
`;

exports[`Component template files Root non ejs files expect .eslintrc.cjs ejsRenderer fn output to match snapshot 1`] = `
"module.exports = {
	\\"extends\\": \\"origami-component\\"
};
"
`;

exports[`Component template files Root non ejs files expect .gitignore ejsRenderer fn output to match snapshot 1`] = `
".DS_Store
.env
/.sass-cache/
/bower_components/
/node_modules/
/build/
.idea/
/demos/local
/coverage
"
`;

exports[`Component template files Root non ejs files expect .npmignore ejsRenderer fn output to match snapshot 1`] = `
".env
.gitignore
.npmignore
.eslintrc.cjs
.remarkrc.cjs
.stylelintrc.cjs
.prettierrc.toml
.eslintignore
.stylelintignore
stories/
src/tsx/
.github/
test/
demos/local/

"
`;

exports[`Component template files Root non ejs files expect .prettierrc.toml ejsRenderer fn output to match snapshot 1`] = `
"printWidth = 80
trailingComma = \\"es5\\"
semi = true
singleQuote = true
useTabs = true
bracketSpacing = false
jsxBracketSameLine = true
arrowParens = \\"avoid\\"
"
`;

exports[`Component template files Root non ejs files expect .remarkrc.cjs ejsRenderer fn output to match snapshot 1`] = `
"module.exports.plugins = [require(\\"remark-preset-lint-origami-component\\")]
"
`;

exports[`Component template files Root non ejs files expect .stylelintignore ejsRenderer fn output to match snapshot 1`] = `
"demos/local/**
"
`;

exports[`Component template files Root non ejs files expect .stylelintrc.cjs ejsRenderer fn output to match snapshot 1`] = `
"module.exports = {
    \\"extends\\": \\"stylelint-config-origami-component\\"
};"
`;

exports[`Component template files demos ejs files expect demo.js ejsRenderer fn output to match snapshot 1`] = `
"import '../../main.js';

document.addEventListener(\\"DOMContentLoaded\\", function () {
	document.dispatchEvent(new CustomEvent('o.DOMContentLoaded'));
});
"
`;

exports[`Component template files demos ejs files expect demo.mustache ejsRenderer fn output to match snapshot 1`] = `
"<div class=\\"o-test-name\\" data-o-component=\\"o-test-name\\"></div>
"
`;

exports[`Component template files demos ejs files expect demo.scss ejsRenderer fn output to match snapshot 1`] = `
"@import '../../main';

@include oTestName();
"
`;

exports[`Component template files src and stories ejs files expect <name>.js ejsRenderer fn output to match snapshot 1`] = `
"class TestName {
	/**
		* Class constructor.
		* @param {HTMLElement} [testNameEl] - The component element in the DOM
		* @param {Object} [options={}] - An options object for configuring the component
		*/
	constructor (testNameEl, options) {
		this.testNameEl = testNameEl;
		this.options = Object.assign({}, {
		}, options || TestName.getDataAttributes(testNameEl));
	}
	/**
		* Get the data attributes from the TestNameElement. If the element is being set up
		* declaratively, this method is used to extract the data attributes from the DOM.
		* @param {HTMLElement} testNameEl - The component element in the DOM
		* @returns {Object} An options object which can be used for configuring the component
		*/
	static getDataAttributes (testNameEl) {
		if (!(testNameEl instanceof HTMLElement)) {
			return {};
		}
		return Object.keys(testNameEl.dataset).reduce((options, key) => {
			// Ignore keys which are not in the component's namespace
			if (!key.match(/^oTestName(\\\\w)(\\\\w+)$/)) {
				return options;
			}
			// Build a concise key and get the option value
			const shortKey = key.replace(/^oTestName(\\\\w)(\\\\w+)$/, (m, m1, m2) => m1.toLowerCase() + m2);
			const value = testNameEl.dataset[key];
			// Try parsing the value as JSON, otherwise just set it as a string
			try {
				options[shortKey] = JSON.parse(value.replace(/'/g, '\\"'));
			} catch (error) {
				options[shortKey] = value;
			}
			return options;
		}, {});
	}
	/**
		* Initialise o-test-name component/s.
		* @param {(HTMLElement|String)} rootElement - The root element to intialise the component in, or a CSS selector for the root element
		* @param {Object} [options={}] - An options object for configuring the component
		* @returns {TestName|TestName[]} The newly constructed TestName components
		*/
	static init (rootElement, options) {
		if (!rootElement) {
			rootElement = document.body;
		}
		if (!(rootElement instanceof HTMLElement)) {
			rootElement = document.querySelector(rootElement);
		}
		if (rootElement instanceof HTMLElement && rootElement.matches('[data-o-component=o-test-name]')) {
			return new TestName(rootElement, options);
		}
		return Array.from(rootElement.querySelectorAll('[data-o-component=\\"o-test-name\\"]'), rootEl => new TestName(rootEl, options));
	}
}

export default TestName;
"
`;

exports[`Component template files src and stories ejs files expect <name>.stories.tsx ejsRenderer fn output to match snapshot 1`] = `
"
import {ComponentStory, ComponentMeta} from '@storybook/react';
import {TestName} from '../src/tsx/test-name';
import './test-name.scss';

export default {
	title: 'Components/o-test-name',
	component: TestName,
	parameters: {},
	args: {},
} as ComponentMeta<typeof TestName>;

const TestNameStory = args => <TestName {...args} />;
export const DefaultTestName: ComponentStory<typeof TestName> = TestNameStory.bind(
	{}
);
"
`;

exports[`Component template files src and stories ejs files expect <name>.tsx ejsRenderer fn output to match snapshot 1`] = `
"export function TestName() {
	return (<div className=\\"o-test-name\\"data-o-component=\\"o-test-name\\"></div>)
}
"
`;

exports[`Component template files src and stories ejs files expect <name>.tsx ejsRenderer fn output to match snapshot 2`] = `
"@import '@financial-times/o-fonts/main';
@import '@financial-times/o-normalise/main';
@include oFonts();
@include oNormalise();

@import \\"@financial-times/o-test-name/main\\";
@include oTestName();
"
`;

exports[`Component template files src and stories ejs files expect _brand.scss ejsRenderer fn output to match snapshot 1`] = `
"/// Helper for \`o-brand\` function.
/// @access private
@function _oTestNameGet($variables, $from: null) {
	@return oBrandGet($component: 'o-test-name', $variables: $variables, $from: $from);
}
/// Helper for \`o-brand\` function.
/// @access private
@function _oTestNameSupports($variant) {
	@return oBrandSupportsVariant($component: 'o-test-name', $variant: $variant);
}
"
`;

exports[`Component template files src and stories ejs files expect _variables.scss ejsRenderer fn output to match snapshot 1`] = `
"// Add any global Sass variables for o-test-name here.
"
`;

exports[`Component template files test ejs files expect <name>.test.js ejsRenderer fn output to match snapshot 1`] = `
"/* eslint-env mocha */
import proclaim from 'proclaim';
import sinon from 'sinon/pkg/sinon-esm.js';
import * as fixtures from './helpers/fixtures.js';
import TestName from '../main.js';

describe(\\"TestName\\", () => {
	it('is defined', () => {
		proclaim.equal(typeof TestName, 'function');
	});

	it('has a static init method', () => {
		proclaim.equal(typeof TestName.init, 'function');
	});

	it(\\"should autoinitialize\\", (done) => {
		const initSpy = sinon.spy(TestName, 'init');
		document.dispatchEvent(new CustomEvent('o.DOMContentLoaded'));
		setTimeout(function(){
			proclaim.equal(initSpy.called, true);
			initSpy.restore();
			done();
		}, 100);
	});

	it(\\"should not autoinitialize when the event is not dispatched\\", () => {
		const initSpy = sinon.spy(TestName, 'init');
		proclaim.equal(initSpy.called, false);
	});

	describe(\\"should create a new o-test-name\\", () => {

		beforeEach(() => {
			fixtures.htmlCode();
		});

		afterEach(() => {
			fixtures.reset();
		});

		it(\\"component array when initialized\\", () => {
			const boilerplate = TestName.init();
			proclaim.equal(boilerplate instanceof Array, true);
			proclaim.equal(boilerplate[0] instanceof TestName, true);
		});

		it(\\"single component when initialized with a root element\\", () => {
			const boilerplate = TestName.init('#element');
			proclaim.equal(boilerplate instanceof TestName, true);
		});
	});
});
"
`;

exports[`Component template files test ejs files expect _main.test.scss ejsRenderer fn output to match snapshot 1`] = `
"@include describe('oTestName mixins') {
	@include describe('oTestName') {
		@include it('does something') {
			@include assert() {
				@include output($selector: false) {
					.test-mixin {
						@include oTestName();
					}
				}
				@include expect($selector: false) {
					.test-mixin .o-test-name {
						display: block;
					}
				}
			}
		}
	}
}
"
`;

exports[`Component template files test ejs files expect fixtures.js ejsRenderer fn output to match snapshot 1`] = `
"let sandboxEl;

function createSandbox() {
	if (document.querySelector('.sandbox')) {
		sandboxEl = document.querySelector('.sandbox');
	} else {
		sandboxEl = document.createElement('div');
		sandboxEl.setAttribute('class', 'sandbox');
		document.body.appendChild(sandboxEl);
	}
}

function reset() {
	sandboxEl.innerHTML = '';
}

function insert(html) {
	createSandbox();
	sandboxEl.innerHTML = html;
}

function htmlCode () {
	const html = \`<div>
		<div class=\\"o-test-name\\" data-o-component=\\"o-test-name\\" id=\\"element\\"></div>
	</div>
	\`;
	insert(html);
}

export {
	htmlCode,
	reset
};
"
`;

exports[`Component template files test ejs files expect index.test.scss ejsRenderer fn output to match snapshot 1`] = `
"@import 'true';
@import '../../main';

$_test-environment: true;

@import 'main.test';
"
`;
